kong/plugins/prometheus/exporter.lua

```diff
@@ -168,6 +174,69 @@ local function collect()
                  "/metrics endpoint: ", err)
   end
 
+  -- erase all target/upstream metrics, prevent exposing old metrics
+  metrics.upstream_target_health:reset()
+
+  -- upstream targets accessible?
+  local balancer = require "kong.runloop.balancer"
+  local upstreams_dict = balancer.get_all_upstreams()
+  for upstream_name, upstream_id in pairs(upstreams_dict) do
+    -- based on logic from kong.db.dao.targets
+    local health_info
+    health_info, err = balancer.get_upstream_health(upstream_id)
+    if err then
+      kong.log.err("failed getting upstream health: ", err)
+    end
+
+    if health_info then
+      local status = 0 -- healthchecks is off
+      for target_name, target_info in pairs(health_info) do
+        if target_info ~= nil and target_info.addresses ~= nil and
+          #target_info.addresses > 0 then
+          for _, address in ipairs(target_info.addresses) do
+            if address.health == "HEALTHY" then
+              status = 1
+              break
+            elseif address.health == "UNHEALTHY" then
+              status = 2
+            else
+              kong.log.err("Only HEALTHY and UNHEALTHY target health status are valid.")
+            end
+          end
+        else
+          status = 3 -- dns error
+        end
+        local upstream_target_state_labels = {
+          { upstream_name, target_name, "healthchecks_off" },
+          { upstream_name, target_name, "healthy" },
+          { upstream_name, target_name, "unhealthy" },
+          { upstream_name, target_name, "dns_error" },
+        }
```

Comment:
the current code is really close to a merging state but i'll push the review a little further:)
the idea here is to reuse table much as possible, in this case we do save lot of typing below but these tables are still created for each upstream.
we could actually create them before the for loop at Line 183
```lua
local upstream_target_state_labels = {0, 0, 0}
for upstream_name, upstream_id in pairs(upstreams_dict) do
...
  for target_name, target_info in pairs(health_info) do
      upstream_target_state_labels[1] = upstream_name
      upstream_target_state_labels[2] = target_name
...
```

Reply:
Thanks for your review. I followed your tip!